File: concepts/06_text_binary_and_serialization/01_unicode_strings_and_code_units/lua/test_02_normalization_and_invalid_encoding.lua
-- Common question: how are canonically equivalent and malformed Unicode inputs handled?
-- Inputs: composed/decomposed accents, invalid UTF-8 bytes, and code-point iteration.
-- Observations: byte equality, absent normalization, utf8.len failure position, and iterator rejection.
-- polyglot-family: text_binary_and_serialization
-- polyglot-concept: unicode_strings_and_code_units
-- polyglot-related: languages/lua/standard_library/09_strings_patterns_utf8/
-- polyglot-related+: test_072_utf8_codepoints_and_byte_indices.lua

local t = require("support.assertions")
local composed = "\u{00e9}"
local decomposed = "e\u{0301}"

t.falsey(composed == decomposed)
t.equal(utf8.len(composed), 1)
t.equal(utf8.len(decomposed), 2)

local invalid = "A\255B"
local length, invalid_position = utf8.len(invalid)
t.equal(length, nil)
t.equal(invalid_position, 2)
t.raises(function()
    for _ in utf8.codes(invalid) do end
end, "invalid UTF-8 code")

-- 标准 utf8 库不做 Unicode normalization 或 grapheme segmentation。
t.done()

File: concepts/06_text_binary_and_serialization/01_unicode_strings_and_code_units/ruby/test_02_normalization_and_invalid_encoding.rb
# 共同问题：规范等价文本和非法字节怎样处理。
# 输入：NFC/NFD、UTF-8 非法字节、valid_encoding?、scrub 和 transcoding；观察：显式规范化及错误策略。
# polyglot-family: text_binary_and_serialization
# polyglot-concept: unicode_strings_and_code_units
# polyglot-related: languages/ruby/language/07_strings_regex_and_patterns/
# polyglot-related+: test_050_encoding_transcoding_and_invalid_bytes.rb

require "assertions"

A = PolyglotAssertions

composed = "\u00E9"
decomposed = "e\u0301"
A.falsey(composed == decomposed)
A.equal(composed, decomposed.unicode_normalize(:nfc))
A.equal(decomposed, composed.unicode_normalize(:nfd))

invalid = "\xFF".b.force_encoding(Encoding::UTF_8)
A.falsey(invalid.valid_encoding?)
A.equal("?", invalid.scrub("?"))
A.raises(Encoding::InvalidByteSequenceError) { invalid.encode(Encoding::UTF_16LE) }

A.done
